### PR TITLE
[MRG] Handle nan/masked values Axes.vlines and hlines

### LINKED ISSUE #7406: NaN causes plt.vlines to not scale y limits
I have reduced the error to the following code snippet which I run in a Jupyter notebook. A NaN as the first item in the series causes plt.vlines to not scale the y limits appropriately. The data values are in the thousands, but the ylimits are [-0.06, 0.06].

```python
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sys

print sys.version
print 'pandas', pd.__version__
print 'matplotlib', matplotlib.__version__
df = pd.DataFrame(index = pd.DatetimeIndex(['2016-10-07', '2016-10-14', '2016-10-21', '2016-10-28'], name='Date'),
                  data = [np.nan, -1000, -3000, 4500],
                  columns = ['Amount'])
print df

plt.figure(figsize=(12,3))
plt.vlines(df.index, 0, df['Amount'])
```

```
2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
pandas 0.19.0
matplotlib 1.5.3
            Amount
Date              
2016-10-07     NaN
2016-10-14 -1000.0
2016-10-21 -3000.0
2016-10-28  4500.0
```

- [ ] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...)
1.5.3, python 2.7.12, Linux

- [ ] How did you install Matplotlib and Python (pip, anaconda, from source ...)
conda

- [ ]
matplotlib-bug

### PR DESCRIPTION
**Pull request by phobson**
Closes #7406. 

Also applied fixes to Axes.hlines and wrote tests for both.

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -979,14 +979,14 @@ def hlines(self, y, xmin, xmax, colors='k', linestyles='solid',
 
         verts = [((thisxmin, thisy), (thisxmax, thisy))
                  for thisxmin, thisxmax, thisy in zip(xmin, xmax, y)]
-        coll = mcoll.LineCollection(verts, colors=colors,
+        lines = mcoll.LineCollection(verts, colors=colors,
                                     linestyles=linestyles, label=label)
```
Comment (dopplershift) [inline lib/matplotlib/axes/_axes.py]: Shouldn't this line need to indent to keep things lined up (pep8 and all that...)

Comment (NelleV) [inline lib/matplotlib/axes/_axes.py]: nice catch!

**Comment by efiring** [comment]
This supports nans in the input, but not masked arrays; in most cases we try to support both.  If you want to handle masked arrays, you could run the arguments through a function analogous to `cbook.safe_masked_invalid`, but going in the opposite direction by using `np.ma.filled(x, np.nan)` if x is floating point.

Overall, we have a somewhat confusing mixture of handling missing points via nans (which only works for floating point, and which requires nanmax etc.) and via masked arrays, for which methods work mostly as they do on plain ndarrays.

**Comment by efiring** [comment]
Second thought: for these functions, would it make more sense to simply filter out the bad values at the start?  I don't see any point in pushing the handling of bad values down into the min and max calculations.  Admittedly, I haven't looked closely.

**Comment by phobson** [comment]
@efiring Filtering was my thought when I first went into this. Let me mess around with feeding masked arrays into this function before we merge.

In either case, handling both masked arrays and sequences with invalid (`np.nan`) values in a simple way will take some thinking.

**Comment by phobson** [comment]
With the current implementation in this PR, masked arrays does *not* work as expected.

For example
```python
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

fig3, ax5 = plt.subplots()
x5 = np.ma.masked_equal([2, 4, 6, 8, 10, 12], 8)
ymin5 = np.ma.masked_equal([0, 1, -1, 0, 2, 1], 2)
ymax5 = np.ma.masked_equal([13, 14, 15, 16, 17, 18], 18)
ax5.vlines(x5, ymin5, ymax5, colors='k', linewidth=2)
```

Produces a plot with 5 lines. Only the line where `x5 == 8` is omitted.

I'll get back to this later this weekend.

**Comment by efiring** [comment]
Filtering a single array is easy: `x = np.ma.compressed(cbook.safe_masked_invalid(x))`. 
For vlines, however, one to three arrays have to be filtered together.  This is accomplished by `cbook.delete_masked_points`.  Right after the unit conversion, insert
```python
    x, ymin, ymax = cbook.delete_masked_points(x, ymin, ymax)
```
and I think that will take care of everything.

**Comment by phobson** [comment]
@efiring Great. Thanks for the tip. I'll include that and the example from my previous comment as a test.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -973,20 +975,19 @@ def hlines(self, y, xmin, xmax, colors='k', linestyles='solid',
...
+        lines = mcoll.LineCollection(verts, colors=colors,
+                                     linestyles=linestyles, label=label)
+        self.add_collection(lines, autolim=False)
+        lines.update(kwargs)
 
         if len(y) > 0:
-            minx = min(xmin.min(), xmax.min())
-            maxx = max(xmin.max(), xmax.max())
+            minx = np.min([xmin.min(), xmax.min()])
+            maxx = np.max([xmin.max(), xmax.max()])
```
Comment (dopplershift) [inline lib/matplotlib/axes/_axes.py]: What's the rationale behind the change from using the builtin `min` and `max` to numpy's? If we're talking micro-performance-optimization, the builtins will be faster.

Comment (phobson) [inline lib/matplotlib/axes/_axes.py]: Originally, `vlines` used the builtin `min` and `max`, while `hlines` used a mix of builtin and numpy. I was going for consistency.

To my eyes, the numpy syntax is more familiar and it feels "safer" (not sure if that's true). I wasn't considering performance with those changes. 

My understanding is also that it's faster:

`x = numpy.ranndom.normal(size=3700)`
`%timeit x.max()`
> The slowest run took 53.27 times longer than the fastest. This could mean that an intermediate result is being cached.
> 100000 loops, best of 3: 2.88 µs per loop
where `(2.9 * 55 ~ 160 µs)`

`%timeit max(x)`
> 1000 loops, best of 3: 231 µs per loop

Comment (dopplershift) [inline lib/matplotlib/axes/_axes.py]: I'm referring specifically to this change:
```python
minx = min(xmin.min(), xmax.min())
```
to
```python
minx = np.min([xmin.min(), xmax.min()])
```
and similar for `max()`--not the change from `min(ymin)` to `ymin.min()`, etc. This latter change I agree with. For the former, I dislike using numpy's `min` just to find the minimum of two scalars. You end up creating an extra list, which then descends into numpy and creates an array. On my system, taking the minimum of two scalars takes 10us with numpy, 4us with the builtin. There's also more for my eye to scan across.

But I don't want to bikeshed this to death (6us is not going to have any impact on matplotlib's slowness)--but I'd be interested if we as developers have consensus on what the convention is or should be.

Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: I agree: builtin min is cleaner for comparing two scalars.  As @anntzer has pointed out, however, there is a difference in the way NaN is handled.  Numpy also has `minimum`, which can be used with two scalars and is faster than using `np.min`:
```
In [1]: import numpy as np
In [2]: min(np.nan, 1.1)
nan
In [3]: min(1.1, np.nan)
1.1
In [7]: timeit np.minimum(2.2, 1.1)
The slowest run took 16.35 times longer than the fastest. This could mean that an intermediate result is being cached
1000000 loops, best of 3: 1.13 µs per loop

In [8]: timeit min(2.2, 1.1)
The slowest run took 11.75 times longer than the fastest. This could mean that an intermediate result is being cached
1000000 loops, best of 3: 197 ns per loop

In [9]: timeit np.min([2.2, 1.1])
The slowest run took 260.96 times longer than the fastest. This could mean that an intermediate result is being cached
100000 loops, best of 3: 9.37 µs per loop
```
Conclusion: if there is any possibility of nans, use np.minimum; otherwise, builtin min is fine.

Comment (phobson) [inline lib/matplotlib/axes/_axes.py]: > I dislike using numpy's min just to find the minimum of two scalars.

Ahh that makes sense. I'll make the change. You've convinced me :)

Comment (dopplershift) [inline lib/matplotlib/axes/_axes.py]: Also interesting that the difference between array scalars and python floats is about a factor of 20 on my system (when using built-in `min`).

Comment (phobson) [inline lib/matplotlib/axes/_axes.py]: @efiring since we're using e.g.,  `cbook.delete_masked_points(x, ymin, ymax)` I think the possibility for NaNs has been removed.

**Comment by dopplershift** [comment]
Test failure is a weird linkchecker failure--seems unrelated.

**Comment by QuLogic** [comment]
Link checker failed because the documentation does not exist; the failure is [here](https://travis-ci.org/matplotlib/matplotlib/jobs/173995346#L1152) and appears to be related to this PR.

**Comment by dopplershift** [comment]
Wow, not sure how I missed that.

**Comment by QuLogic** [comment]
We should probably change Travis to not run link checker when the doc build fails.

**Comment by phobson** [comment]
That failure is really confusing to me because the tests pass and the docs build on my local branch.

Nonetheless, I think the latest commit (deleting masked/invalid points after we make sure the first arg is an iterable) should do the trick.

**Comment by dopplershift** [comment]
All green--I think we're good here.

**Comment by dopplershift** [comment]
If this is actually intended for 2.0, it will need to be cherry-picked.

**Comment by efiring** [comment]
back-ported to v2.x as ad192f0

**Comment by phobson** [comment]
thanks for the guidance and merging everyone
